File: DSA Code/Linked List/1. Singly Linked List.cpp
#include<iostream>
using namespace std;

class Node
{
    public:
    int data;
    Node *next = nullptr;
    Node(int data)
    {
        this->data = data;
    }
};

class SLL
{
    Node *head = nullptr;
    public:
    SLL()
    { }

    bool isEmpty()
    {
        if(head == nullptr)
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    void insertAtStart(int data)
    {
        if(head == nullptr)
        {
            head = new Node(data);
        }
        else
        {
            Node *temp = new Node(data);
            temp->next = head;
            head = temp;
        }
    }

    void insertAtLast(int data)
    {
        if(head == nullptr)
        {
            head = new Node(data);
        }
        else
        {
            Node *temp = head;
            while(temp->next != nullptr)
            {
                temp = temp->next;
            }
            temp->next = new Node(data);
        }
    }

    // helper function
    Node* searchNode(int key)
    {
        Node *temp = head;
        while(temp != nullptr)
        {
            if(temp->data == key)
            {
                return temp;
            }
            temp = temp->next;
        }
        return nullptr;
    }

    void insertAfter(int key, int data)
    {
        Node *searchedNode = searchNode(key);
        if(searchedNode != nullptr)
        {
            if(searchedNode->next == nullptr)
            {
                searchedNode->next = new Node(data);
            }
            else
            {
                Node *temp = new Node(data);
                temp->next = searchedNode->next;
                searchedNode->next = temp;
            }
        }
        else
        {
            cout<<"Node not found!" <<endl;
        }
    }

    void show()
    {
        cout<<"Linked List: ";
        if(head != nullptr)
        {
            Node *temp = head;
            while(temp != nullptr)
            {
                cout<<temp->data <<"  ";
                temp = temp->next;
            }
        }
        cout<<endl <<"---------------------------------------" <<endl;
    }

    void deleteFirst()
    {
        if(head != nullptr)
        {
            if(head->next == nullptr)
            {
                delete head;
                head = nullptr;
            }
            else
            {
                Node *temp = head;
                head = head->next;
                delete temp;
            }
        }
        else
        {
            cout<<"Linked List is already empty!" <<endl;
        }
    }

    void deleteLast()
    {
        if(head != nullptr)
        {
            if(head->next == nullptr)
            {
                delete head;
                head = nullptr;
            }
            else
            {
                Node *temp = head;
                while(temp->next->next != nullptr)
                {
                    temp = temp->next;
                }
                Node *delNode = temp->next;
                temp->next = nullptr;
                delete delNode;
            }
        }
        else
        {
            cout<<"Linked List is already empty!" <<endl;
        }
    }

    void deletenode(int key)
    {
        Node *delNode = searchNode(key);
        if(delNode != nullptr)
        {
            // only one node
            if(head->next == nullptr)
            {
                delete head;
                head = nullptr;
            }
            // first node but having more than one node
            else if(delNode == head && head->next != nullptr)
            {
                Node *temp = head;
                head = head->next;
                delete temp;
            }
            // middle and last node
            else
            {
                Node *temp = head;
                while(temp->next != delNode)
                {
                    temp = temp->next;
                }
                temp->next = delNode->next;
                delete delNode;
            }
        }
        else
        {
            cout<<"Node doesn't Exist!" <<endl;
        }
    }

    void deleteSLL()
    {
        if(head != nullptr)
        {
            Node *temp = head;
            while(head != nullptr)
            {
                head = head->next;
                delete temp;
                temp = head;
            }
        }
    }

    ~SLL()
    {
        deleteSLL();
    }
};

int main()
{
    SLL list;
    list.insertAtStart(20);
    list.insertAtLast(30);
    list.insertAtLast(50);
    list.insertAtStart(10);
    list.insertAfter(50,60);
    list.insertAfter(30,40);
    list.show();

    list.deleteFirst();
    list.deleteLast();
    list.show();

    list.deletenode(30);
    list.show();

    list.deletenode(40);
    list.show();

    list.deletenode(20);
    list.show();

    list.deletenode(50);
    list.show();

    return 0;
}

File: JS Basic DSA/Linked List/1. Singly Linked List.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class SLL {
    #head;
    constructor() {
        this.head = null;
    }

    isEmpty() {
        return this.head === null;
    }

    insertAtStart(data) {
        if (this.head === null) {
            this.head = new Node(data);
        } else {
            let temp = new Node(data);
            temp.next = this.head;
            this.head = temp;
        }
    }

    insertAtLast(data) {
        if (this.head === null) {
            this.head = new Node(data);
        } else {
            let temp = this.head;
            while (temp.next !== null) {
                temp = temp.next;
            }
            temp.next = new Node(data);
        }
    }

    // helper function
    searchNode(key) {
        let temp = this.head;
        while (temp !== null) {
            if (temp.data === key) {
                return temp;
            }
            temp = temp.next;
        }
        return null;
    }

    insertAfter(key, data) {
        let searchedNode = this.searchNode(key);
        if (searchedNode !== null) {
            if (searchedNode.next === null) {
                searchedNode.next = new Node(data);
            } else {
                let temp = new Node(data);
                temp.next = searchedNode.next;
                searchedNode.next = temp;
            }
        } else {
            console.log("Node not found!");
        }
    }

    show() {
        let temp = this.head;
        let result = "Linked List: ";
        while (temp !== null) {
            result += temp.data + "  ";
            temp = temp.next;
        }
        console.log(result);
        console.log("---------------------------------------");
    }

    deleteFirst() {
        if (this.head !== null) {
            if (this.head.next === null) {
                // delete only node
                this.head = null;
            } else {
                // delete first node
                this.head = this.head.next;
            }
        } else {
            console.log("Linked List is already empty!");
        }
    }

    deleteLast() {
        if (this.head !== null) {
            if (this.head.next === null) {
                // delete only node
                this.head = null;
            } else {
                // delete last node
                let temp = this.head;
                while (temp.next.next !== null) {
                    temp = temp.next;
                }
                temp.next = null;
            }
        } else {
            console.log("Linked List is already empty!");
        }
    }

    deleteNode(key) {
        let delNode = this.searchNode(key);

        if (delNode !== null) {
            // first node
            if (delNode === this.head) {
                this.head = this.head.next;
            }
            // deleting any middle or last node
            else {
                let temp = this.head;
                while (temp.next !== delNode) {
                    temp = temp.next;
                }
                temp.next = delNode.next;
            }
        } else {
            console.log("Node doesn't Exist!");
        }
    }

    deleteSLL() {
        // JS garbage collector handles deletion
        this.head = null;
    }
}


let list = new SLL();

list.insertAtStart(20);
list.insertAtLast(30);
list.insertAtLast(50);
list.insertAtStart(10);
list.insertAfter(50, 60);
list.insertAfter(30, 40);
list.show();

list.deleteFirst();
list.deleteLast();
list.show();

list.deleteNode(30);
list.show();

list.deleteNode(40);
list.show();

list.deleteNode(20);
list.show();

list.deleteNode(50);
list.show();